keps/sig-node/20190517-image-decryption.md

```diff
@@ -0,0 +1,330 @@
+
+
+An unspecified `nil` or empty `""` ImageDecryptSecret is equivalent to the backwards-compatible
+default behavior as if the ImageDecryptSecret feature is disabled.
+
+#### Examples
+
+Suppose we operate a cluster that lets users create a secret of type `ImageDecryptSecret`. 
+
+For the kubectl command line, we propose that the user create a secret of type `image-decrypt` and give it a name. One or multiple private keys may then be stored under this secret.
+
+```bash
+kubectl create secret image-decrypt <secret name> --decrypt-secret=<private key>[:<password>] [--decrypt-secret=<private key>[:<password>]]
+```
+`private key` - base64 representation of key
+
+`password` (optional) - password in *cleartext* if the given private key is protected with the password
+
+For example,
+```bash
+kubectl create secret image-decrypt keysecret --decrypt-secret="base64OfPrivateKey":"password"
+```
+See it in action in the [demo](https://youtu.be/S3FK4y5McOk?t=19)
+
+A variant for passing the ImageEncrypSecret is the following:
+
+```bash
+kubectl create secret image-decrypt <secret name> --from-file=<path to file>:[:<password>] 
+```
+
+`path to file` - The path of the private key in binary format
+
+`password` (optional) - password in *cleartext* if the given private key is protected with the password
+For example,
+```bash
+kubectl create secret image-decrypt keysecret --from-file="/path/to/private_key":"password"
+```
+
+
+Then when a user creates a workload, they can choose the desired ImageDecryptSecret to use 
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+  labels:
+    name: nginx
+spec:
+  containers:
+  - name: nginx
+    image: localhost:5000/nginx:enc
+    imageDecryptSecrets:
+    - name: <secret name>
+    imagePullPolicy: Always
+    ports:
+    - containerPort: 80
+
+```
+
+ImageDecryptSecrets can be added to the `service account` by,
+
+```bash
+kubectl patch serviceaccount <account name> -p '{"imageDecryptSecrets":[{"name":<secret name>, "images":[<image name>}]}'
+```
+or while creating a secret account,
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  creationTimestamp: 2015-08-07T22:02:39Z
+  name: default
+  namespace: default
+  selfLink: /api/v1/namespaces/default/serviceaccounts/default
+  uid: 052fb0f4-3d50-11e5-b066-42010af0d7b6
+secrets:
+- name: default-token-uudge
+imagePullSecrets:
+- name: myregistrykey
+imageDecryptSecrets:
+- name: <secret name>
+- image: [<image name>]
+```
+
+For example,
+```bash
+kubectl patch serviceaccount default -p '{"imageDecryptSecrets":[{"name":"keysecret", "images":["localhost:5000/nginx:enc"]}]}'
+```
+
+#### Image Handler
+
+The privake keys are extracted from `ImageDecryptSecret` and passed to the CRI as by bundling them in `DecryptParams` as a part of `PullImageRequest`:
+
+
+```protobuf
+// DecryptParams is used to send decryption parameters to the image service of the CRI
+message DecryptParams {
+    repeated string privateKeyPasswds = 1;
```

Comment:
(See also the question about the format of `SecretTypeDecryptKey`.)

So the format is “passwords”? Is that multiple passwords in a single `string` value, as the name of the field seems to indicate? If so, how are they separated/formatted/escaped? Or is it just supposed to be a set of strings, one password per string?

How does this relate to the
```bash
kubectl create secret image-decrypt keysecret --decrypt-secret="base64OfPrivateKey":"password"```
```
example above? That seems to suggest that the secret is actually some kind of non-password private key.

Reply:
it's just supposed to be a set of strings with optional password. So it could look like,

```golang
[]string{"private_key1:passswd", "private_key2:passwd"}
```

Reply:
Why wouldn’t that be a well-typed Go data structure / protobuf message with two separate fields instead of using an in-line `:` separator?

And what is the format of `private_key` in the above? Examples suggest a base64-encoded string…
- Is the base64 encoding essential? Wouldn’t it be natural to use `bytes` in protobuf without the base64 overhead?
- What is the format of the contents of the base64-encoded bytes?
- If it is a set, how does the consumer know which key to use for what? (I guess the keys are not raw cryptographic keys, but something with an in-line ID?)

---

Maybe all of this does not necessarily have to be explicitly defined in _this_ API, it could just point to a separate spec / implementation (maintained by OCI?), but ultimately the format must be documented _somewhere_ if the secrets are supposed to be interoperable.

Compare most of the other `SecretType`* constants, which either explicitly describe the fields (like `SecretTypeBasicAuth`), or refer to other established formats (like `SecretTypeDockerConfigJson`); or how `AuthConfig` is granular at this point.

Reply:
Yeah, I agree that we should change from base64 of the private keys to a well-typed Go data structure/protobuf message to represent private keys and the corresponding passwords. I will update the KEP document followed by the PR for actual code. 

In case there is a set of keys, they are all passed to a runtime like containerd/cri-o. It's runtime's job to decide what to do with these keys and the given layers. The way it currently works in containerd is, it just interates over the list of keys to see which one is able to unwrap the symmetric key needed for actual decryption of the layer blob.

Reply:
Agree that the field needs to be better defined.

All questions listed in https://github.com/kubernetes/enhancements/pull/1066/files#r315240393 are good questions. :)

Reply:
@mtrmac I have removed the dependency on the base64 encoding. 

@mtrmac @Random-Liu I have modified [API](https://github.com/kubernetes/enhancements/blob/edc643b6a55a2ed4cc0231447352bc3ef65136e7/keps/sig-node/20190517-image-decryption.md#api) section to define the fields better. Please let me know what do you think.